libraries/AP_NavEKF/AP_NavEKF_Source.cpp

```diff
@@ -0,0 +1,431 @@
+        // check velz
+        switch ((SourceZ)_source[i].velz.get()) {
+        case SourceZ::NONE:
+            break;
+        case SourceZ::GPS:
+            gps_required = true;
+            break;
+        case SourceZ::EXTNAV:
+            visualodom_required = true;
+            break;
+        case SourceZ::BARO:
+        case SourceZ::RANGEFINDER:
+        case SourceZ::BEACON:
+        default:
+            // invalid velz value
+            hal.util->snprintf(failure_msg, failure_msg_len, "Check EK3_SRC_VELZ%s", idx_str[i]);
+            return false;
+        }
+
+        // check yaw
+        switch ((SourceYaw)_source[i].yaw.get()) {
+        case SourceYaw::NONE:
+        case SourceYaw::COMPASS:
+        case SourceYaw::EXTERNAL:
+        case SourceYaw::EXTERNAL_COMPASS_FALLBACK:
+            // valid yaw value
+            break;
+        default:
+            // invalid posz value
+            hal.util->snprintf(failure_msg, failure_msg_len, "Check EK3_SRC_YAW%s", idx_str[i]);
+            return false;
+        }
+    }
+
+    // check all required sensor are available
+    const char* ekf_requires_msg = "EK3_SRC requires %s";
+    if (baro_required && (dal.baro().num_instances() == 0)) {
+        hal.util->snprintf(failure_msg, failure_msg_len, ekf_requires_msg, "Baro");
+        return false;
+    }
+
+    if (beacon_required && (dal.beacon() == nullptr || !dal.beacon()->enabled())) {
+        hal.util->snprintf(failure_msg, failure_msg_len, ekf_requires_msg, "Beacon");
+        return false;
+    }
+
+    if (compass_required && dal.compass().get_num_enabled() == 0) {
+        hal.util->snprintf(failure_msg, failure_msg_len, ekf_requires_msg, "Compass");
+        return false;
+    }
+
+    if (gps_required && (dal.gps().num_sensors() == 0)) {
+        hal.util->snprintf(failure_msg, failure_msg_len, ekf_requires_msg, "GPS");
+        return false;
+    }
+
+    if (optflow_required && !dal.opticalflow_enabled()) {
+        hal.util->snprintf(failure_msg, failure_msg_len, ekf_requires_msg, "OpticalFlow");
+        return false;
+    }
+
+    if (rangefinder_required && (dal.rangefinder() == nullptr || !dal.rangefinder()->has_orientation(ROTATION_PITCH_270))) {
+        hal.util->snprintf(failure_msg, failure_msg_len, ekf_requires_msg, "RangeFinder");
+        return false;
+    }
+
+    if (visualodom_required) {
+        bool visualodom_available = false;
+#if HAL_VISUALODOM_ENABLED
+        auto *vo = AP::dal().visualodom();
+        visualodom_available = vo && vo->enabled();
+#endif
+        if (!visualodom_available) {
+            hal.util->snprintf(failure_msg, failure_msg_len, ekf_requires_msg, "VisualOdom");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+// get source by id
+AP_NavEKF_Source::SourceXY AP_NavEKF_Source::getPosXYSourceByIndex(uint8_t idx) const
+{
+    if (idx >= AP_NAKEKF_SOURCE_COUNT) {
+        return SourceXY::NONE;
+    }
+    return (SourceXY)_source[idx].posxy.get();
```

Comment:
This is because we've got a single enum for the XY axis vs separating it into two (one for position, one for velocity).  I think this is part of the question above so I'll mark this issue as "resolved" but it will be changed if we decide to split the enums.